## Assignment Preparation

Pandas provides the `read_csv()` function to read data stored as a csv file into a Pandas `DataFrame`:

In [52]:
import pandas as pd
titanic = pd.read_csv("titanic.csv")

Pandas supports many different file formats or data sources (csv, Excel, SQL, JSON, ...), each of them with the prefix `read_*`.

Make sure to always check the data after reading it in. When displaying a `DataFrame`, the first and last 5 rows will be shown by default:

In [53]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

To get the first N entries of a `DataFrame`, use the `head()` method with the required number of rows as the argument:

In [4]:
titanic.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Pandas also provides a `tail()` method. `titanic.tail(10)` will return the last 10 rows of the `DataFrame`:

In [5]:
titanic.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

The method `info()` provides technical information about a `DataFrame`:

In [6]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


The output tells us:

- We have a `DataFrame`
- There are 891 entries (or rows)
- Each row has a row label (aka the index) with values ranging from 0 to 890
- The table has 12 columns
    - Most columns have a value for each of the rows (all 891 values are non-null)
    - Some columns do have missing values and less than 891 non-null values
- The columns `Name`, `Sex`, `Cabin`, and `Embarked` consist of textual data (strings, aka `object`). The other columns are numerical data with some of them whole numbers (aka `int64`) and others are real numbers (aka `float64`)
- The kind of data in the different columns is summarized by listing the `dtype`s
- The approximate amount of memory used to hold the `DataFrame` is provided as well

`DataFrame.shape` is an attribute containing the number of rows and columns: `(nrows, ncolumns)`

In [54]:
titanic.shape

(891, 12)

In [50]:
titanic.shape[0] # nrows

891

In [55]:
titanic.shape[1] # ncolumns

12

If we want to calculate a given statistic (e.g., mean, median, sum, min, max) for each category in a column (e.g., male/female in the `Sex` column), we can use the `groupby()` method. This method works by:

1. Splitting the data into groups
2. Applying a function to each group independently
3. Combining the results into a data structure

The apply and combine steps are typically done together in Pandas. For example, if we want to calculate the mean for each numerical column in our dataset grouped by `Sex`, we would do the following:

In [8]:
titanic.groupby("Sex").mean()

PassengerId  Survived    Pclass        Age     SibSp     Parch  \
Sex                                                                      
female   431.028662  0.742038  2.159236  27.915709  0.694268  0.649682   
male     454.147314  0.188908  2.389948  30.726645  0.429809  0.235702   

             Fare  
Sex                
female  44.479818  
male    25.523893

It doesn't make much sense to get the average value of some of our columns, such as `Pclass`. If we are only interested in the average age for each gender, the selection of columns with `[]` is supported as well:

In [9]:
titanic.groupby("Sex")["Age"].mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

Grouping can be done by multiple columns at the same time, just provide the column names as a list to the `groupby()` method. For example, what is the mean ticket fare price for each of the `Sex` and `Pclass` (cabin class) combinations?

In [10]:
titanic.groupby(["Sex", "Pclass"])["Fare"].mean()

Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64

In [56]:
titanic.groupby("Sex")[["Age", "Fare"]].mean()

Age       Fare
Sex                         
female  27.915709  44.479818
male    30.726645  25.523893

We can also sort our data according to a specific column. For example, let's sort the `titanic` data according to the age of the passengers:

In [58]:
titanic.sort_values(by="Age").head()

PassengerId  Survived  Pclass                             Name     Sex  \
803          804         1       3  Thomas, Master. Assad Alexander    male   
755          756         1       2        Hamalainen, Master. Viljo    male   
644          645         1       3           Baclini, Miss. Eugenie  female   
469          470         1       3    Baclini, Miss. Helene Barbara  female   
78            79         1       2    Caldwell, Master. Alden Gates    male   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
803  0.42      0      1    2625   8.5167   NaN        C  
755  0.67      1      1  250649  14.5000   NaN        S  
644  0.75      2      1    2666  19.2583   NaN        C  
469  0.75      2      1    2666  19.2583   NaN        C  
78   0.83      0      2  248738  29.0000   NaN        S

Or sort the `titanic` data according to the cabin class and age in descending order:

In [13]:
titanic.sort_values(by=['Pclass','Age'],ascending=False).head()

PassengerId  Survived  Pclass                       Name     Sex   Age  \
851          852         0       3        Svensson, Mr. Johan    male  74.0   
116          117         0       3       Connors, Mr. Patrick    male  70.5   
280          281         0       3           Duane, Mr. Frank    male  65.0   
483          484         1       3     Turkula, Mrs. (Hedwig)  female  63.0   
326          327         0       3  Nysveen, Mr. Johan Hansen    male  61.0   

     SibSp  Parch  Ticket    Fare Cabin Embarked  
851      0      0  347060  7.7750   NaN        S  
116      0      0  370369  7.7500   NaN        Q  
280      0      0  336439  7.7500   NaN        Q  
483      0      0    4134  9.5875   NaN        S  
326      0      0  345364  6.2375   NaN        S

We can count the number of records for each category in a column with the `value_counts()` method. For example: let's get the number of passengers in each of the cabin classes:

In [14]:
titanic["Pclass"].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [61]:
titanic[["Sex", "Pclass"]].value_counts()

Sex     Pclass
male    3         347
female  3         144
male    1         122
        2         108
female  1          94
        2          76
dtype: int64

The function is a shortcut, as it is actually a `groupby()` operation in combination with the counting of the number of records within each group:

In [15]:
titanic.groupby("Pclass")["Pclass"].count()

Pclass
1    216
2    184
3    491
Name: Pclass, dtype: int64

In [60]:
titanic.groupby(["Sex", "Pclass"])["Pclass"].count()

Sex     Pclass
female  1          94
        2          76
        3         144
male    1         122
        2         108
        3         347
Name: Pclass, dtype: int64

To get basic statistics, use the `describe()` method, which provides a quick overview of the numerical data in a `DataFrame`:

In [72]:
titanic.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [68]:
titanic[["Age", "Fare"]].describe()

Age        Fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

In [78]:
titanic["Age"].describe()['mean':'min']

mean    29.699118
std     14.526497
min      0.420000
Name: Age, dtype: float64

If we wanted to include all columns, not just numerical ones, we could include the optional parameter `include="all"`:

In [17]:
titanic.describe(include="all")

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891     2   
top             NaN         NaN         NaN  Braund, Mr. Owen Harris  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383838    2.308642                      NaN   NaN   
std      257.353842    0.486592    0.836071                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

               Age       SibSp       Parch  Ticket        Fare    Cabin  \
count   714.000000  891.000000  891.000000     891  891.000000      204   
unique         NaN         NaN         NaN     681         NaN      147   
top            NaN         NaN         NaN  347082         NaN  B96 B98   
freq           NaN         NaN         NaN       7         NaN        4   
mean     29.699118    0.523008    0.381594     NaN   32.204208      NaN   
std      14.526497    1.102743    0.806057     NaN   49.693429      NaN   
min       0.420000    0.000000    0.000000     NaN    0.000000      NaN   
25%      20.125000    0.000000    0.000000     NaN    7.910400      NaN   
50%      28.000000    0.000000    0.000000     NaN   14.454200      NaN   
75%      38.000000    1.000000    0.000000     NaN   31.000000      NaN   
max      80.000000    8.000000    6.000000     NaN  512.329200      NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

Instead of the predefined statistics, specific combinations of aggregating statistics for given columns can be defined using the `agg()` method:

In [21]:
titanic.agg({"Age":['min','max','median','sum'],
             "Fare":['min','max','median','mean']})

Age        Fare
min         0.42    0.000000
max        80.00  512.329200
median     28.00   14.454200
sum     21205.17         NaN
mean         NaN   32.204208

We can apply functions over rows of data using the `apply()` method.

Syntax: `DataFrame.apply(function)`

For example:

In [80]:
# create a function
def sex_to_num(x):
    if x == 'male':
        return 1
    if x == 'female':
        return 0

# apply the function to the Sex column and create a new column
titanic['Sex_N'] = titanic["Sex"].apply(sex_to_num)

titanic[["Sex","Sex_N"]].head()

Sex  Sex_N
0    male      1
1  female      0
2  female      0
3  female      0
4    male      1

In [81]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Sex_N  
0        0         A/5 21171   7.2500   NaN        S      1  
1        0          PC 17599  71.2833   C85        C      0  
2        0  STON/O2. 3101282   7.9250   NaN        S      0  
3        0            113803  53.1000  C123        S      0  
4        0            373450   8.0500   NaN        S      1  
..     ...               ...      ...   ...      ...    ...  
886      0            211536  13.0000   NaN        S      1  
887      0            112053  30.0000   B42        S      0  
888      2        W./C. 6607  23.4500   NaN        S      0  
889      0            111369  30.0000  C148        C      1  
890      0            370376   7.7500   NaN        Q      1  

[891 rows x 13 columns]

If we want to get the index location of a value in a column, we can use the `idxmax()` method. For example, if we wanted to get the passenger with the longest name on the Titanic, we could find its index location by:

In [33]:
titanic["Name"].str.len().idxmax()

307

By using the `str.len()` method, the function is applied to each of the names individually. With this information, we are able to utilize the `loc` operator to determine what the longest name is:

In [34]:
titanic.loc[titanic["Name"].str.len().idxmax(), "Name"]

'Penasco y Castellana, Mrs. Victor de Satode (Maria Josefa Perez de Soto y Vallejo)'

We can filter specific rows from a `DataFrame` using our `[]` notation. For example, if we wanted to get all entries where `Age` is greater than 35, we could write:

In [18]:
titanic[titanic["Age"] > 35].head()

PassengerId  Survived  Pclass  \
1             2         1       1   
6             7         0       1   
11           12         1       1   
13           14         0       3   
15           16         1       2   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
6       0     17463  51.8625   E46        S  
11      0    113783  26.5500  C103        S  
13      5    347082  31.2750   NaN        S  
15      0    248706  16.0000   NaN        S

In [82]:
titanic["Age"] > 35

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

To select rows based on a conditional expression, we use a condition inside the `[]`. The condition inside the `[]` `titanic['Age'] > 35` checks for which rows the `Age` column has a value larger than 35. 

The output of the conditional expression (>, ==, !=, <, <=, >=, ...) is actually a Pandas `Series` of boolean values (`True` or `False`) with the same number of rows as the original `DataFrame`. Such a `Series` of boolean values can be used to filter the `DataFrame` by putting it in between the `[]`. Only rows for which the value is `True` will be selected.

If we want to combine multiple conditional statements, each condition must be surrounded by parentheses. Moreover, we cannot use `or` or `and`, but we need to use the operators `|` and `&`, respectively:

In [19]:
titanic[(titanic["Age"] > 35) | (titanic["Fare"] < 600)].head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

To select multiple columns, use a list of column names within the `[]`. The inner square brackets define a Python list with column names, whereas the outer brackets are used to select the data from a Pandas `DataFrame`:

In [20]:
titanic[["Age","Sex"]].head()

Age     Sex
0  22.0    male
1  38.0  female
2  26.0  female
3  35.0  female
4  35.0    male

We can append two `DataFrame`s together using the `append()` method. 

Syntax: `DataFrame.append(DataFrame)`

For example, if we wanted to append our area/population `DataFrame` to our Titanic data:

In [36]:
area = pd.Series({'California': 423967, 'Texas': 695662,
                  'New York': 141297, 'Florida': 170312,
                  'Illinois': 149995})
population = pd.Series({'California': 38332521, 'Texas': 26448193,
                 'New York': 19651127, 'Florida': 19552860,
                 'Illinois': 12882135})
data = pd.DataFrame({'area':area, 'pop':population})
data

area       pop
California  423967  38332521
Texas       695662  26448193
New York    141297  19651127
Florida     170312  19552860
Illinois    149995  12882135

In [39]:
titanic.append(data)

/var/folders/vy/r2k1d_qd7qq72mp6t471xvq80000gp/T/ipykernel_5842/733124079.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  titanic.append(data)


PassengerId  Survived  Pclass  \
0                   1.0       0.0     3.0   
1                   2.0       1.0     1.0   
2                   3.0       1.0     3.0   
3                   4.0       1.0     1.0   
4                   5.0       0.0     3.0   
...                 ...       ...     ...   
California          NaN       NaN     NaN   
Texas               NaN       NaN     NaN   
New York            NaN       NaN     NaN   
Florida             NaN       NaN     NaN   
Illinois            NaN       NaN     NaN   

                                                         Name     Sex   Age  \
0                                     Braund, Mr. Owen Harris    male  22.0   
1           Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
2                                      Heikkinen, Miss. Laina  female  26.0   
3                Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
4                                    Allen, Mr. William Henry    male  35.0   
...                                                       ...     ...   ...   
California                                                NaN     NaN   NaN   
Texas                                                     NaN     NaN   NaN   
New York                                                  NaN     NaN   NaN   
Florida                                                   NaN     NaN   NaN   
Illinois                                                  NaN     NaN   NaN   

            SibSp  Parch            Ticket     Fare Cabin Embarked  Sex_N  \
0             1.0    0.0         A/5 21171   7.2500   NaN        S    1.0   
1             1.0    0.0          PC 17599  71.2833   C85        C    0.0   
2             0.0    0.0  STON/O2. 3101282   7.9250   NaN        S    0.0   
3             1.0    0.0            113803  53.1000  C123        S    0.0   
4             0.0    0.0            373450   8.0500   NaN        S    1.0   
...           ...    ...               ...      ...   ...      ...    ...   
California    NaN    NaN               NaN      NaN   NaN      NaN    NaN   
Texas         NaN    NaN               NaN      NaN   NaN      NaN    NaN   
New York      NaN    NaN               NaN      NaN   NaN      NaN    NaN   
Florida       NaN    NaN               NaN      NaN   NaN      NaN    NaN   
Illinois      NaN    NaN               NaN      NaN   NaN      NaN    NaN   

                area         pop  
0                NaN         NaN  
1                NaN         NaN  
2                NaN         NaN  
3                NaN         NaN  
4                NaN         NaN  
...              ...         ...  
California  423967.0  38332521.0  
Texas       695662.0  26448193.0  
New York    141297.0  19651127.0  
Florida     170312.0  19552860.0  
Illinois    149995.0  12882135.0  

[896 rows x 15 columns]

Additionally, we can use the `concat()` function, which performs concatenation operations of multiple tables along one of the axes (row-wise or column-wise). By default, concatenation is along axis 0, so the resulting table combines the rows of the input tables. 

For example, if we wanted to combine our two previous datasets together along the rows:

In [40]:
pd.concat([titanic, data])

PassengerId  Survived  Pclass  \
0                   1.0       0.0     3.0   
1                   2.0       1.0     1.0   
2                   3.0       1.0     3.0   
3                   4.0       1.0     1.0   
4                   5.0       0.0     3.0   
...                 ...       ...     ...   
California          NaN       NaN     NaN   
Texas               NaN       NaN     NaN   
New York            NaN       NaN     NaN   
Florida             NaN       NaN     NaN   
Illinois            NaN       NaN     NaN   

                                                         Name     Sex   Age  \
0                                     Braund, Mr. Owen Harris    male  22.0   
1           Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
2                                      Heikkinen, Miss. Laina  female  26.0   
3                Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
4                                    Allen, Mr. William Henry    male  35.0   
...                                                       ...     ...   ...   
California                                                NaN     NaN   NaN   
Texas                                                     NaN     NaN   NaN   
New York                                                  NaN     NaN   NaN   
Florida                                                   NaN     NaN   NaN   
Illinois                                                  NaN     NaN   NaN   

            SibSp  Parch            Ticket     Fare Cabin Embarked  Sex_N  \
0             1.0    0.0         A/5 21171   7.2500   NaN        S    1.0   
1             1.0    0.0          PC 17599  71.2833   C85        C    0.0   
2             0.0    0.0  STON/O2. 3101282   7.9250   NaN        S    0.0   
3             1.0    0.0            113803  53.1000  C123        S    0.0   
4             0.0    0.0            373450   8.0500   NaN        S    1.0   
...           ...    ...               ...      ...   ...      ...    ...   
California    NaN    NaN               NaN      NaN   NaN      NaN    NaN   
Texas         NaN    NaN               NaN      NaN   NaN      NaN    NaN   
New York      NaN    NaN               NaN      NaN   NaN      NaN    NaN   
Florida       NaN    NaN               NaN      NaN   NaN      NaN    NaN   
Illinois      NaN    NaN               NaN      NaN   NaN      NaN    NaN   

                area         pop  
0                NaN         NaN  
1                NaN         NaN  
2                NaN         NaN  
3                NaN         NaN  
4                NaN         NaN  
...              ...         ...  
California  423967.0  38332521.0  
Texas       695662.0  26448193.0  
New York    141297.0  19651127.0  
Florida     170312.0  19552860.0  
Illinois    149995.0  12882135.0  

[896 rows x 15 columns]

We can also merge datasets together using the `merge()` function. This differs from the `append()` method and the `concat()` function as the `merge()` function need a common column, which is used as a key to combine the information.

Syntax: `pd.merge(DataFrame1, DataFrame2, how, on)`

Parameters:

- `how`: specifies how to merge; default: `"inner"`
    - Other options include  `"left"`, `"right"`, and `"outer"`
- `on`: specifies on what level to do the merging

For example:

In [93]:
data1 = {"name": ["Sally", "Mary", "John"],
         "age": [50, 40, 30]}

data2 = {"name": ["Sally", "Peter", "Micky"],
         "age": [77, 44, 22]}

df1 = pd.DataFrame(data1)
df2 = pd.DataFrame(data2)

newdf = pd.merge(df1, df2, how='left', on='name')

newdf.head()

name  age_x  age_y
0  Sally     50   77.0
1   Mary     40    NaN
2   John     30    NaN

We can rename our row and column labels using the `rename()` method. The argument to pass is a dictionary with the keys as the current names and the values as the new names to update the corresponding names. 

For example:

In [45]:
df2.head()

name  age
0  Sally   77
1  Peter   44
2  Micky   22

In [95]:
df2.rename(columns={"name":"Name", "age":"Age"})

Name  Age
0  Sally   77
1  Peter   44
2  Micky   22

The function `reset_index()` can be used to convert any level of an index to a column, e.g.,:

In [107]:
df2.reset_index()

index   name  age
0      0  Sally   77
1      1  Peter   44
2      2  Micky   22

We can return all entries without `Null` values with the `notna()` method. It returns a `True` for each row the values are not a `Null` value. As such, this can be combined with `[]` to filter the data table:

In [49]:
titanic[titanic["Age"].notna()]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Sex_N  
0        0         A/5 21171   7.2500   NaN        S      1  
1        0          PC 17599  71.2833   C85        C      0  
2        0  STON/O2. 3101282   7.9250   NaN        S      0  
3        0            113803  53.1000  C123        S      0  
4        0            373450   8.0500   NaN        S      1  
..     ...               ...      ...   ...      ...    ...  
885      5            382652  29.1250   NaN        Q      0  
886      0            211536  13.0000   NaN        S      1  
887      0            112053  30.0000   B42        S      0  
889      0            111369  30.0000  C148        C      1  
890      0            370376   7.7500   NaN        Q      1  

[714 rows x 13 columns]

In [108]:
titanic["Age"].notna()

0       True
1       True
2       True
3       True
4       True
       ...  
886     True
887     True
888    False
889     True
890     True
Name: Age, Length: 891, dtype: bool

In [109]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
 12  Sex_N        891 non-null    int64  
dtypes: float64(2), int64(6), object(5)
memory usage: 90.6+ KB
